open_spiel/algorithms/infostate_cfr.h

```diff
@@ -0,0 +1,156 @@
+// Copyright 2019 DeepMind Technologies Ltd. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef OPEN_SPIEL_ALGORITHMS_INFOSTATE_CFR_H_
+#define OPEN_SPIEL_ALGORITHMS_INFOSTATE_CFR_H_
+
+#include <map>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "open_spiel/algorithms/infostate_tree.h"
+#include "open_spiel/algorithms/cfr.h"
+#include "open_spiel/policy.h"
+#include "open_spiel/spiel.h"
+#include "open_spiel/spiel_utils.h"
+
+// This file provides a vectorized implementation of CFR.
+// This is intended for advanced usage and the code may not be as readable
+// as a more basic algorithm. See cfr.h for a basic implementation.
+//
+// This code uses a preconstructed infostate trees of each player. It updates
+// all infostates at a tree depth at once. While the current implementation is
+// okay in efficiency (~10x faster than CFRSolver), it could be further
+// improved:
+//
+// - Use contiguous memory blocks for storing regrets/current/cumul
+//   for each action.
+// - Skip subtrees that have zero reach probabilities.
+// - Use stack allocation: most trees are small. If they exceed the allocated
+//   limit (1024? nodes), use something like a linked list with these large
+//   memory blocks.
+//
+// More todos:
+// - Implement CFR+, Predictive CFR+, as local regret minimizers.
+// - Provide custom leaf evaluation (neural net)
+//
+// If you decide to make contributions to this code, please open up an issue
+// on github first. Thank you!
+
+namespace open_spiel {
+namespace algorithms {
+
+// A helper struct that allows to propagate reach probs / cf values
+// up and down the tree.
+struct InfostateTreeValuePropagator {
+  // Tree and the tree structure information. These must not change!
```

Comment:
This seems ugly. Shouldn't these values be parameters to the propagation function rather than members here?